src/world_generation/world_generator.rs

```diff
@@ -21,63 +22,108 @@ impl WorldGenerator {
-                        }
-
-                        let height = biome_config.terrain_offset
-                            + biome_config.terrain_scale
-                                * noise.at([i as f32 * s + x as f32, j as f32 * s + z as f32]);
-
-                        let cube_height = height.floor() as i32;
-
-                        for y in 0..cube_height {
-                            let block_at_height = biome_config.get_block_at(cube_height - y - 1);
-
-                            if let Some(block) = block_at_height {
-                                chunk.add_cube(
-                                    Vector3::new(
-                                        i as f32 * s + x as f32,
-                                        y as f32,
-                                        j as f32 * s + z as f32,
-                                    ),
-                                    block,
-                                    0,
-                                );
-                            }
-                        }
-                    }
-                }
+
+                let (chunk, mut new_trees) = WorldGenerator::create_random_chunk(seed, [x0, z0], &mut noise);
 
                 chunks.push(chunk);
+                possible_tree_pos.append(&mut new_trees);
             }
         }
 
+        for tree_pos in possible_tree_pos {
+            // Find in which chunk the tree is
+            let tree_x = tree_pos[0];
+            let tree_y = tree_pos[1];
+            let tree_z = tree_pos[2];
+
+            let chunk_corner_x = (tree_x as f32 / s).floor() as i32;
+            let chunk_corner_z = (tree_z as f32 / s).floor() as i32;
+            
+            // Find the exact from the chunk
+            let width = 2 * n_chunks + 1;
+            let i_shifted = chunk_corner_x + n_chunks;
+            let j_shifted = chunk_corner_z + n_chunks;
+            let index = i_shifted * width + j_shifted;
+
+            TreeGenerator::try_and_plant_a_tree(Vector3::new(
+                tree_x as f32,
+                tree_y as f32,
+                tree_z as f32
+            ), &mut chunks[index as usize]);
+
+        }
+
         World::new(chunks)
     }
 
+    // Returns the chunk with the block inside and a list of possible trees in that chunk
+    pub fn create_random_chunk(seed: u64, chunk_coord: [f32; 2], noise: &mut MultiscalePerlinNoise) -> (Chunk, Vec<[i32; 3]>) {
+        let x0 = chunk_coord[0];
+        let z0 = chunk_coord[1];
+
+        let mut chunk = Chunk::new([x0, z0]);
+        let mut possible_tree_pos_in_chunk: Vec<[i32; 3]> = vec![];
+
+        // get the height from the perlin noise for each block
+        for x in 0..8 {
+            for z in 0..8 {
+                // Get biome
+                let biome_t: u64 = BiomeGenerator::find_closest_biome(seed, x + x0 as i32, z + z0 as i32);
+                
+                let biome_config = &BIOMES[biome_t as usize];
```

Comment:
That's not good. It's actually seriously wrong. This is rust and not C. We can talk together to find something better than this.

Reply:
The reason why I consider this bad is **once and once more the "encapsulation principle":** the caller of a function (*or of a class, or anything that you use...*) must not be aware of the internal logic of the function to use the output of the function. 

Here it's not the case, you need to know that you receive a "type" (`biome_t`) but that actually this type is not a "rust type" but it is the index in an external array of many biomes... 

So we won't accept this. 

At the very least, the function `find_closest_biome` returns `&BiomeConfig`. I'm not even sure that this is good, but it is at least better.